src/lexitools/commands/correspondences.py

```diff
@@ -1,161 +1,373 @@
             concept_dict = concepticon.conceptlists[concept_list].concepts
             self.concepts_subset = {c.concepticon_id for c in
                                     concept_dict.values() if
                                     c.concepticon_id}
-        else:
-            self.concepts_subset = set(
-                concepts.values())  # All observed concepts
+        else:  # All observed concepts
+            self.concepts_subset = set(concepts.values())
 
-        self.genera_to_lang = defaultdict(set)
         self.lang_to_concept = defaultdict(set)
-        self.tokens = defaultdict(list)
-        self.genera_to_family = {}
+        self.data = defaultdict(lambda:defaultdict(list))
 
         for row in progressbar(db.iter_table('FormTable'),
                                desc="Loading data..."):
-
-            lang_id, family = langs[row["Language_ID"]]
-            # skip if it has no glottocode
-            # or no recognized glottocode,
-            # or no phonological information
-            if lang_id is None \
-                    or lang_id not in self.lang_to_genera \
-                    or row["Segments"] is None \
-                    or "".join(
-                [x for x in row["Segments"] if x is not None]) == "":
+            concept = concepts[row["Parameter_ID"]]
+            if concept not in self.concepts_subset or \
+                    row["Language_ID"] not in langs or row["Segments"] is None:
                 continue
 
-            concept = concepts[row["Parameter_ID"]]
-            genus = self.lang_to_genera[lang_id]
-            token = list(self._iter_phonemes(row["Segments"]))
-
-            if concept in self.concepts_subset:
-                self.genera_to_lang[genus].add(lang_id)
-                self.lang_to_concept[lang_id].add(concept)
-                self.tokens[(lang_id, concept)].append(token)
-                if genus not in self.genera_to_family:
-                    self.genera_to_family[genus] = family
-        self.concepts_intersection = []
-        log.info(r"Total number of concepts kept: {}".format(
-            len(self.concepts_subset)))
-
-    def _iter_phonemes(self, segments):
+            try:
+                token = list(self._iter_phonemes(row))
+            except ValueError: continue  # unknown sounds
+            if token == [""]: continue
+
+            syllables = len(lingpy.sequence.sound_classes.syllabify(token,
+                                                                    output="nested"))
+            lang = langs[row["Language_ID"]]
+            row = Row(lang=lang, syllables=syllables,
+                      token=token, concept=concept, id=row["ID"],
+                      original_token=" ".join(row["Segments"]), dataset=row["dataset"])
+
+            # TODO: add cognates
+            # TODO: ignore Loans (row["Loan"] exists at least in some datasets)
+            self.data[(lang, concept)][" ".join(token)].append(row)
```

Comment:
yeah, remove entries with row['Loan']. Something like `if len(row.get('Loan', "")` should catch them